### PR TITLE
Duplicated branch rules

### PR DESCRIPTION
**Pull request by vilchik-elena**

### CODE REVIEW

File: javascript-checks/src/main/java/org/sonar/javascript/checks/AbstractDuplicateBranchImplementationCheck.java [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+  @Override
+  public void visitSwitchStatement(SwitchStatementTree tree) {
+    boolean hasClauseWithoutJump = false;
+    boolean hasDefaultCase = false;
+
+    List<SwitchClauseTree> branches = new ArrayList<>();
+
+    for (int i = 0; i < tree.cases().size(); i++) {
+      SwitchClauseTree caseTree = tree.cases().get(i);
+      boolean isLast = (i == tree.cases().size() - 1);
```
Comment (SonarTech) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/AbstractDuplicateBranchImplementationCheck.java]: MINOR Remove those useless parentheses. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUselessParenthesesCheck)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

File: javascript-checks/src/main/java/org/sonar/javascript/checks/AbstractDuplicateBranchImplementationCheck.java (hunk 1 above)
Comment (inverno) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/AbstractDuplicateBranchImplementationCheck.java]: I like them actually

File: javascript-frontend/src/main/java/org/sonar/javascript/metrics/LineVisitor.java [hunk 2]
```diff
@@ -19,41 +19,33 @@
...
 import org.sonar.plugins.javascript.api.tree.lexical.SyntaxToken;
+import org.sonar.plugins.javascript.api.visitors.DoubleDispatchVisitor;
 
 /**
  * Visitor that computes the number of lines of code of a file.
  */
-public class LineVisitor extends SubscriptionVisitor {
+public class LineVisitor extends DoubleDispatchVisitor {
 
   private Set<Integer> lines = Sets.newHashSet();
```
Comment (inverno) [inline javascript-frontend/src/main/java/org/sonar/javascript/metrics/LineVisitor.java]: We should stop using `Sets.newHashSet` and switch to `new HashSet<>();`

File: javascript-checks/src/main/java/org/sonar/javascript/checks/AbstractDuplicateBranchImplementationCheck.java (hunk 1 above)
Comment (vilchik-elena) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/AbstractDuplicateBranchImplementationCheck.java]: @inverno yeah, i've kept them for the same reason. In fact this rule should be soften soon, so no problem

File: javascript-checks/src/test/resources/checks/duplicateBranchImplementation.js [hunk 3]
```diff
@@ -1,98 +1,241 @@
...
-}
+  }else {
+    first();
+    second();
+  }
+
+  if (a) {
+    foo();
+    bar();
+  }
```
Comment (inverno) [inline javascript-checks/src/test/resources/checks/duplicateBranchImplementation.js]: @vilchik-elena what's this for?

File: javascript-frontend/src/main/java/org/sonar/javascript/metrics/LineVisitor.java (hunk 2 above)
Comment (inverno) [inline javascript-frontend/src/main/java/org/sonar/javascript/metrics/LineVisitor.java]: Done on my side

File: javascript-checks/src/test/resources/checks/duplicateBranchImplementation.js (hunk 3 above)
Comment (inverno) [inline javascript-checks/src/test/resources/checks/duplicateBranchImplementation.js]: Added comment to explain

File: javascript-checks/src/main/java/org/sonar/javascript/checks/AbstractDuplicateBranchImplementationCheck.java (hunk 1 above)
Comment (inverno) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/AbstractDuplicateBranchImplementationCheck.java]: Yep, I know, but I wanted to bully @SonarTech a bit :smile:

File: javascript-checks/src/main/java/org/sonar/javascript/checks/AbstractDuplicateBranchImplementationCheck.java [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+  @Override
+  public void visitSwitchStatement(SwitchStatementTree tree) {
+    boolean hasFallthrough = false;
+    boolean hasDefaultCase = false;
+
+    List<Tree> branches = new ArrayList<>();
+
+    for (int i = 0; i < tree.cases().size(); i++) {
+      SwitchClauseTree caseTree = tree.cases().get(i);
+      boolean isLast = (i == tree.cases().size() - 1);
```
Comment (SonarTech) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/AbstractDuplicateBranchImplementationCheck.java]: MINOR Remove those useless parentheses. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUselessParenthesesCheck)

**Pull request merged**